File: packages/sdk/src/events.ts
/**
 * Fixed Soroban event topic prefixes emitted by SubscriptionVault.
 * Use with RPC `getEvents` topic filters (Symbol ScVals).
 *
 * Topics layout:
 * - created:   ["sub", "created", subscription_id, subscriber, merchant]
 * - paid:      ["sub", "paid", subscription_id, subscriber, merchant]
 * - cancelled: ["sub", "cancelled", subscription_id, subscriber, merchant]
 */
export const EVENT_TOPICS = {
  PREFIX: "sub",
  CREATED: "created",
  PAID: "paid",
  CANCELLED: "cancelled",
} as const;

export type VaultEventKind =
  | typeof EVENT_TOPICS.CREATED
  | typeof EVENT_TOPICS.PAID
  | typeof EVENT_TOPICS.CANCELLED;

/** Decoded `sub/created` event data map fields. */
export interface SubscriptionCreatedData {
  token: string;
  amount: bigint;
  intervalSecs: bigint;
}

/** Decoded `sub/paid` event data map fields. */
export interface PaymentProcessedData {
  token: string;
  amount: bigint;
  billedAt: bigint;
}

/** Topic filter fragments for RPC `getEvents` (wildcard-friendly). */
export function vaultEventTopicFilter(
  kind: VaultEventKind,
): [string, string, "*", "*", "*"] {
  return [EVENT_TOPICS.PREFIX, kind, "*", "*", "*"];
}

File: contracts/subscription_vault/src/events.rs
//! Contract events for indexers, keepers, and merchant dashboards.
//!
//! Fixed topic prefixes (`sub` + action) make RPC `getEvents` filters cheap.
//! Dynamic `#[topic]` fields (`subscription_id`, addresses) are indexed too.

use soroban_sdk::{contractevent, Address};

/// Emitted when a subscriber opens a new recurring bill.
#[contractevent(topics = ["sub", "created"])]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SubscriptionCreated {
    #[topic]
    pub subscription_id: u64,
    #[topic]
    pub subscriber: Address,
    #[topic]
    pub merchant: Address,
    pub token: Address,
    pub amount: i128,
    pub interval_secs: u64,
}

/// Emitted after a successful allowance pull for one billing cycle.
#[contractevent(topics = ["sub", "paid"])]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PaymentProcessed {
    #[topic]
    pub subscription_id: u64,
    #[topic]
    pub subscriber: Address,
    #[topic]
    pub merchant: Address,
    pub token: Address,
    pub amount: i128,
    /// Ledger unix timestamp written to `Subscription.last_billed`.
    pub billed_at: u64,
}

/// Emitted when the subscriber deactivates a subscription.
#[contractevent(topics = ["sub", "cancelled"])]
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SubscriptionCancelled {
    #[topic]
    pub subscription_id: u64,
    #[topic]
    pub subscriber: Address,
    #[topic]
    pub merchant: Address,
}
